{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Full data is inside data.7z file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read DRKG data as pandas data frame\n",
    "# takes around 9 seconds\n",
    "import pandas as pd\n",
    "drkg_df = pd.read_csv('data/drkg/drkg_no_hetionet_dgidb.tsv', sep='\\t')\n",
    "triplets = drkg_df.values.tolist()\n",
    "# print(drkg_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_edge(e, src_type, tgt_type):\n",
    "    edge_arr = e.split('::')\n",
    "    edge_type = edge_arr[1].lower().replace(' ', '_')\n",
    "    str2name = {'ddi-interactor-in': 'ddi_interactor',\n",
    "                'x-atc': 'belongs2ATC', 'a+': 'agonism_activation',\n",
    "                'a-': 'antagonism_blocking', 'b': 'binding', 'c': 'inhibit_cell_growth', 'd': 'drug_targets',\n",
    "                'e+': 'increase_expression_production', 'e-': 'decrease_expression_production',\n",
    "                'e': 'affect_expression_production', 'g': 'promote_progression', 'h': 'same_protein_or_complex',\n",
    "                'i': 'signaling_pathway', 'j': 'role_in_pathogenesis', 'k': 'metabolism_pharmacokinetics',\n",
    "                'l': 'improper_regulation', 'md': 'diognostic_biomarker', 'mp': 'disease_progress_biomarker',\n",
    "                'n': 'inhibit', 'o': 'transport_channel', 'pa': 'alleviate_reduce', 'pr': 'prevent_suppress', 'q': 'production_by_cell_population',\n",
    "                'rg': 'regulate', 'sa': 'sideEffect_adverseEvent', 't': 'treatment', 'te': 'possible_therapeutic_effect',\n",
    "                'u': 'causal_mutation', 'ud': 'mutations_affecting_disease_course', 'v+': 'activate', 'w': 'enhance_response', 'x': 'overexpression',\n",
    "                'y': 'polymorphisms_alter_risk', 'z': 'enzyme_activity', 'ptmod': 'post_translational_modification',\n",
    "                'humgenhumgen:gene:gene': 'hum_gen_hum_gen', 'virgenhumgen:gene:gene': 'vir_gen_hum_gen', 'drugvirgen:compound:gene': 'drug_vir_gen',\n",
    "                'drughumgen:compound:gene': 'drug_hum_gen', 'activation': 'activate', 'target': 'gene_target'}\n",
    "    dir = f'_{src_type}2{tgt_type}'\n",
    "    if edge_type in str2name:\n",
    "        return edge_arr[0] + '_' + str2name[edge_type] + dir\n",
    "    return edge_arr[0] + '_' + edge_type + dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_bulks = {} # dictionary of src_type > edge_type > tgt_type > List of 3 tuples (src_id, tgt_id, attributes)\n",
    "node_bulks = {} # dictionary of node_type > List of 2 tuples (id, attributes)\n",
    "\n",
    "def insert2edge_bulks(src_type, edge_type, tgt_type, src_id, tgt_id, attr):\n",
    "    if not src_type in edge_bulks:\n",
    "        edge_bulks[src_type] = {}\n",
    "    if not edge_type in edge_bulks[src_type]:\n",
    "        edge_bulks[src_type][edge_type] = {}\n",
    "    if not tgt_type in edge_bulks[src_type][edge_type]:\n",
    "        edge_bulks[src_type][edge_type][tgt_type] = []\n",
    "    edge_bulks[src_type][edge_type][tgt_type].append(\n",
    "        (src_id, tgt_id, attr))\n",
    "\n",
    "def insert2node_bulks(node_type, id, attr):\n",
    "    if not node_type in node_bulks:\n",
    "        node_bulks[node_type] = []\n",
    "    node_bulks[node_type].append((id, attr))\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process Schema info from DRKG (takes around 45 seconds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_types = {} # dictionary of string to string to dictionary\n",
    "node_types = {} # dictionary of string to dictionary\n",
    "for tri in triplets:\n",
    "    [src, edge, tgt] = tri\n",
    "    src_arr = src.split('::')\n",
    "    src_type = src_arr[0].replace(' ', '_')\n",
    "    src_id = str(src_arr[1])\n",
    "    tgt_arr = tgt.split('::')\n",
    "    tgt_type = tgt_arr[0].replace(' ', '_')\n",
    "    tgt_id = str(tgt_arr[1])\n",
    "\n",
    "    if not src_type in node_types:\n",
    "        node_types[src_type] = {}\n",
    "    if not tgt_type in node_types:\n",
    "        node_types[tgt_type] = {}\n",
    "    \n",
    "    # add the edge type per type couple\n",
    "    type_edge = f'{src_type}::{tgt_type}'\n",
    "    if not type_edge in edge_types:\n",
    "        edge_types[type_edge] = {}\n",
    "    edge = parse_edge(edge, src_type, tgt_type)\n",
    "        \n",
    "    if not edge in edge_types[type_edge]:\n",
    "        edge_types[type_edge][edge] = {}\n",
    "    \n",
    "    # insert to edge bulks (no need to insert node bulks since node will be created automatically)\n",
    "    insert2edge_bulks(src_type, edge, tgt_type, src_id, tgt_id, {})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def py2TigerType(o):\n",
    "    s = str(type(o))\n",
    "    if 'int' in s:\n",
    "        return 'INT'\n",
    "    if 'float' in s:\n",
    "        return 'FLOAT'\n",
    "    if 'bool' in s:\n",
    "        return 'BOOL'\n",
    "    return 'STRING'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse hetionet data to create schema\n",
    "# takes around 21 seconds\n",
    "import json\n",
    "hetionet =  json.load(open('data/hetionet/hetionet-v1.0.json/hetionet-v1.0.json'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse hetionet nodes\n",
    "for node in hetionet['nodes']:\n",
    "    kind = node['kind']\n",
    "    kind = kind.replace(' ', '_')\n",
    "    if kind not in node_types:\n",
    "        node_types[kind] = {}\n",
    "    \n",
    "    attr = {}\n",
    "    for prop in node:\n",
    "        if prop == 'kind' or prop == 'identifier':\n",
    "            continue\n",
    "        if 'data' in prop:\n",
    "            for d in node['data']:\n",
    "                node_types[kind][d] = py2TigerType(node['data'][d])\n",
    "                if isinstance(node['data'][d], list):\n",
    "                    attr[d] = ','.join(node['data'][d])\n",
    "                else:\n",
    "                    attr[d] = node['data'][d]\n",
    "                \n",
    "        else:\n",
    "            node_types[kind][prop] = py2TigerType(node[prop])\n",
    "            attr[prop] = node[prop]\n",
    "    insert2node_bulks(kind, node['identifier'], attr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse hetionet edges\n",
    "# takes around 58 seconds\n",
    "for edge in hetionet['edges']:\n",
    "    src_type = edge['source_id'][0].replace(' ', '_')\n",
    "    tgt_type = edge['target_id'][0].replace(' ', '_')\n",
    "    kind = 'HETIONET_' + edge['kind'] + f'_{src_type}2{tgt_type}'\n",
    "    type_edge = f'{src_type}::{tgt_type}'\n",
    "    \n",
    "    if type_edge not in edge_types:\n",
    "        edge_types[type_edge] = {}\n",
    "    if kind not in edge_types[type_edge]:\n",
    "        edge_types[type_edge][kind] = {}\n",
    "    attr = {}\n",
    "    for prop in edge:\n",
    "        if prop in {'kind', 'identifier', 'source_id', 'target_id'}:\n",
    "            continue\n",
    "        if 'data' in prop:\n",
    "            for d in edge['data']:\n",
    "                edge_types[type_edge][kind][d] = py2TigerType(edge['data'][d])\n",
    "                if isinstance(edge['data'][d], list):\n",
    "                    attr[d] = ','.join(edge['data'][d])\n",
    "                else:\n",
    "                    attr[d] = edge['data'][d]\n",
    "        else:\n",
    "            edge_types[type_edge][kind][prop] = py2TigerType(edge[prop])\n",
    "            if isinstance(edge[prop], list):\n",
    "                attr[prop] = ','.join(edge[prop])\n",
    "            else:\n",
    "                attr[prop] = edge[prop]\n",
    "\n",
    "    insert2edge_bulks(src_type, kind, tgt_type, edge['source_id'][1], edge['target_id'][1], attr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add node types from DGIDB manually, takes around 13 seconds\n",
    "node_types['Gene']['claim_name'] = 'STRING'\n",
    "node_types['Compound']['drug_claim_name'] = 'STRING'\n",
    "node_types['Compound']['drug_claim_primary_name'] = 'STRING'\n",
    "node_types['Compound']['drug_name'] = 'STRING'\n",
    "\n",
    "dgidb_df = pd.read_csv('data/dgidb/interactions.tsv', sep='\\t')\n",
    "\n",
    "# add edge types and properties\n",
    "dir = 'Compound::Gene'\n",
    "if dir not in edge_types:\n",
    "    edge_types[dir] = {}\n",
    "dgidb_df = dgidb_df[dgidb_df['entrez_id'].notnull()\n",
    "                    & dgidb_df['drug_concept_id'].notnull()\n",
    "                    & dgidb_df['interaction_types'].notnull()]\n",
    "for i, row in dgidb_df.iterrows():\n",
    "    edge_name = 'DGIDB_' + row['interaction_types'].replace(' ', '_').replace(\n",
    "        ',', '_') + '_Compound2Gene'\n",
    "\n",
    "    if edge_name not in edge_types[dir]:\n",
    "        edge_types[dir][edge_name] = {}\n",
    "    row = row.fillna('')\n",
    "    edge_types[dir][edge_name]['interaction_claim_source'] = 'STRING'\n",
    "    edge_types[dir][edge_name]['interaction_types'] = 'STRING'\n",
    "    edge_types[dir][edge_name]['interaction_group_score'] = 'FLOAT'\n",
    "    edge_types[dir][edge_name]['PMIDs'] = 'STRING'\n",
    "    edge = {\n",
    "        'interaction_claim_source': row['interaction_claim_source'],\n",
    "        'interaction_types': row['interaction_types'],\n",
    "        'interaction_group_score': row['interaction_group_score'],\n",
    "        'PMIDs': row['PMIDs']\n",
    "    }\n",
    "    comp_id = row['drug_concept_id'].split(':')[1]\n",
    "    gene_id = str(int(row['entrez_id']))\n",
    "    insert2edge_bulks('Compound', edge_name, 'Gene', comp_id, gene_id, edge)\n",
    "    comp_attr = {\n",
    "        'drug_claim_name': row['drug_claim_name'],\n",
    "        'drug_claim_primary_name': row['drug_claim_primary_name'],\n",
    "        'drug_name': row['drug_name']\n",
    "    }\n",
    "    gene_attr = {'name': row['gene_name'], 'claim_name': row['gene_claim_name']}\n",
    "    insert2node_bulks('Compound', comp_id, comp_attr)\n",
    "    insert2node_bulks('Gene', gene_id, gene_attr)\n",
    "\n",
    "# np.sum(dgidb_df['entrez_id'].notnull() & dgidb_df['drug_concept_id'].notnull())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'{\"Gene\": {\"name\": \"STRING\", \"description\": \"STRING\", \"source\": \"STRING\", \"license\": \"STRING\", \"url\": \"STRING\", \"chromosome\": \"STRING\", \"claim_name\": \"STRING\"}, \"Compound\": {\"name\": \"STRING\", \"license\": \"STRING\", \"source\": \"STRING\", \"inchikey\": \"STRING\", \"inchi\": \"STRING\", \"url\": \"STRING\", \"drug_claim_name\": \"STRING\", \"drug_claim_primary_name\": \"STRING\", \"drug_name\": \"STRING\"}, \"Disease\": {\"name\": \"STRING\", \"source\": \"STRING\", \"license\": \"STRING\", \"url\": \"STRING\"}, \"Atc\": {}, \"Tax\": {}, \"Molecular_Function\": {\"name\": \"STRING\", \"source\": \"STRING\", \"license\": \"STRING\", \"url\": \"STRING\"}, \"Side_Effect\": {\"name\": \"STRING\", \"source\": \"STRING\", \"license\": \"STRING\", \"url\": \"STRING\"}, \"Biological_Process\": {\"name\": \"STRING\", \"source\": \"STRING\", \"license\": \"STRING\", \"url\": \"STRING\"}, \"Pathway\": {\"name\": \"STRING\", \"license\": \"STRING\", \"source\": \"STRING\", \"url\": \"STRING\"}, \"Anatomy\": {\"name\": \"STRING\", \"source\": \"STRING\", \"license\": \"STRING\", \"url\": \"STRING\", \"mesh_id\": \"STRING\", \"bto_id\": \"STRING\"}, \"Cellular_Component\": {\"name\": \"STRING\", \"source\": \"STRING\", \"license\": \"STRING\", \"url\": \"STRING\"}, \"Symptom\": {\"name\": \"STRING\", \"source\": \"STRING\", \"license\": \"STRING\", \"url\": \"STRING\"}, \"Pharmacologic_Class\": {\"name\": \"STRING\", \"class_type\": \"STRING\", \"source\": \"STRING\", \"license\": \"STRING\", \"url\": \"STRING\"}}'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "json.dumps(node_types)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def attr_dict2schema(d):\n",
    "    if len(d) < 1:\n",
    "        return ''\n",
    "    s = ''\n",
    "    l = []\n",
    "    for k,v in d.items():\n",
    "        l.append(f'{k} {v}')\n",
    "    return ', ' + ', '.join(l)\n",
    "# create schema\n",
    "schema = ''\n",
    "for t in node_types:\n",
    "    schema += f'CREATE VERTEX {t} (PRIMARY_ID Id STRING{attr_dict2schema(node_types[t])}) With primary_id_as_attribute=\"true\"\\n'\n",
    "for endpoints in edge_types:\n",
    "    [source_name, target_name] = endpoints.split('::')\n",
    "    for edge_name in edge_types[endpoints]:\n",
    "        schema += f'CREATE DIRECTED EDGE {edge_name} (FROM {source_name}, TO {target_name}{attr_dict2schema(edge_types[endpoints][edge_name])})\\n'\n",
    "# print(schema)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Connect to TigerGraph Cloud instance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Hello GSQL'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pyTigerGraph as tg \n",
    "host = 'https://derman.i.tgcloud.io'\n",
    "# host = 'http://localhost'\n",
    "secret = \"xxxx\"\n",
    "graph_name = \"derman\"\n",
    "user_name = \"tigergraph\"\n",
    "# password = \"tigergraph\"\n",
    "password = \"123456\"\n",
    "# no need auth on local instance\n",
    "token = tg.TigerGraphConnection(host=host, graphname=graph_name, username=user_name, password=password).getToken(secret, \"1000000\")[0]\n",
    "conn = tg.TigerGraphConnection(host=host, graphname=graph_name, username=user_name, password=password, apiToken=token)\n",
    "# conn = tg.TigerGraphConnection(host=host, graphname=graph_name, username=user_name, password=password)\n",
    "conn.echo()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Execute gsql to create schema in tgcloud"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully created vertex types: [Gene].\n",
      "Successfully created vertex types: [Compound].\n",
      "Successfully created vertex types: [Disease].\n",
      "Successfully created vertex types: [Atc].\n",
      "Successfully created vertex types: [Tax].\n",
      "Successfully created vertex types: [Molecular_Function].\n",
      "Successfully created vertex types: [Side_Effect].\n",
      "Successfully created vertex types: [Biological_Process].\n",
      "Successfully created vertex types: [Pathway].\n",
      "Successfully created vertex types: [Anatomy].\n",
      "Successfully created vertex types: [Cellular_Component].\n",
      "Successfully created vertex types: [Symptom].\n",
      "Successfully created vertex types: [Pharmacologic_Class].\n",
      "Successfully created edge types: [bioarx_hum_gen_hum_gen_Gene2Gene].\n",
      "Successfully created edge types: [bioarx_vir_gen_hum_gen_Gene2Gene].\n",
      "Successfully created edge types: [GNBR_activate_Gene2Gene].\n",
      "Successfully created edge types: [GNBR_production_by_cell_population_Gene2Gene].\n",
      "Successfully created edge types: [GNBR_regulate_Gene2Gene].\n",
      "Successfully created edge types: [GNBR_binding_Gene2Gene].\n",
      "Successfully created edge types: [GNBR_signaling_pathway_Gene2Gene].\n",
      "Successfully created edge types: [GNBR_increase_expression_production_Gene2Gene].\n",
      "Successfully created edge types: [GNBR_same_protein_or_complex_Gene2Gene].\n",
      "Successfully created edge types: [GNBR_enhance_response_Gene2Gene].\n",
      "Successfully created edge types: [GNBR_affect_expression_production_Gene2Gene].\n",
      "Successfully created edge types: [INTACT_physical_association_Gene2Gene].\n",
      "Successfully created edge types: [INTACT_association_Gene2Gene].\n",
      "Successfully created edge types: [INTACT_colocalization_Gene2Gene].\n",
      "Successfully created edge types: [INTACT_dephosphorylation_reaction_Gene2Gene].\n",
      "Successfully created edge types: [INTACT_cleavage_reaction_Gene2Gene].\n",
      "Successfully created edge types: [INTACT_direct_interaction_Gene2Gene].\n",
      "Successfully created edge types: [INTACT_phosphorylation_reaction_Gene2Gene].\n",
      "Successfully created edge types: [INTACT_adp_ribosylation_reaction_Gene2Gene].\n",
      "Successfully created edge types: [INTACT_ubiquitination_reaction_Gene2Gene].\n",
      "Successfully created edge types: [INTACT_protein_cleavage_Gene2Gene].\n",
      "Successfully created edge types: [STRING_reaction_Gene2Gene].\n",
      "Successfully created edge types: [STRING_catalysis_Gene2Gene].\n",
      "Successfully created edge types: [STRING_activate_Gene2Gene].\n",
      "Successfully created edge types: [STRING_inhibition_Gene2Gene].\n",
      "Successfully created edge types: [STRING_other_Gene2Gene].\n",
      "Successfully created edge types: [STRING_binding_Gene2Gene].\n",
      "Successfully created edge types: [STRING_post_translational_modification_Gene2Gene].\n",
      "Successfully created edge types: [STRING_expression_Gene2Gene].\n",
      "Successfully created edge types: [HETIONET_interacts_Gene2Gene].\n",
      "Successfully created edge types: [HETIONET_regulates_Gene2Gene].\n",
      "Successfully created edge types: [HETIONET_covaries_Gene2Gene].\n",
      "Successfully created edge types: [bioarx_drug_vir_gen_Compound2Gene].\n",
      "Successfully created edge types: [bioarx_drug_hum_gen_Compound2Gene].\n",
      "Successfully created edge types: [DRUGBANK_gene_target_Compound2Gene].\n",
      "Successfully created edge types: [DRUGBANK_enzyme_Compound2Gene].\n",
      "Successfully created edge types: [DRUGBANK_carrier_Compound2Gene].\n",
      "Successfully created edge types: [GNBR_affect_expression_production_Compound2Gene].\n",
      "Successfully created edge types: [GNBR_agonism_activation_Compound2Gene].\n",
      "Successfully created edge types: [GNBR_inhibit_Compound2Gene].\n",
      "Successfully created edge types: [GNBR_metabolism_pharmacokinetics_Compound2Gene].\n",
      "Successfully created edge types: [GNBR_antagonism_blocking_Compound2Gene].\n",
      "Successfully created edge types: [GNBR_increase_expression_production_Compound2Gene].\n",
      "Successfully created edge types: [GNBR_binding_Compound2Gene].\n",
      "Successfully created edge types: [GNBR_decrease_expression_production_Compound2Gene].\n",
      "Successfully created edge types: [GNBR_transport_channel_Compound2Gene].\n",
      "Successfully created edge types: [GNBR_enzyme_activity_Compound2Gene].\n",
      "Successfully created edge types: [INTACT_physical_association_Compound2Gene].\n",
      "Successfully created edge types: [INTACT_direct_interaction_Compound2Gene].\n",
      "Successfully created edge types: [INTACT_association_Compound2Gene].\n",
      "Successfully created edge types: [HETIONET_upregulates_Compound2Gene].\n",
      "Successfully created edge types: [HETIONET_binds_Compound2Gene].\n",
      "Successfully created edge types: [HETIONET_downregulates_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_inhibitor_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_agonist_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_modulator_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_allosteric_modulator_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_blocker_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_antagonist_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_positive_modulator_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_antibody_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_activator_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_inverse_agonist_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_partial_agonist_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_agonist_antagonist_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_vaccine_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_negative_modulator_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_antagonist_allosteric_modulator_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_inducer_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_antisense_oligonucleotide_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_agonist_allosteric_modulator_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_inhibitory_allosteric_modulator_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_suppressor_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_blocker_activator_Compound2Gene].\n",
      "Successfully created edge types: [DGIDB_blocker_inhibitor_Compound2Gene].\n",
      "Successfully created edge types: [bioarx_covid2_acc_host_gene_Disease2Gene].\n",
      "Successfully created edge types: [bioarx_coronavirus_ass_host_gene_Disease2Gene].\n",
      "Successfully created edge types: [HETIONET_associates_Disease2Gene].\n",
      "Successfully created edge types: [HETIONET_upregulates_Disease2Gene].\n",
      "Successfully created edge types: [HETIONET_downregulates_Disease2Gene].\n",
      "Successfully created edge types: [DRUGBANK_belongs2ATC_Compound2Atc].\n",
      "Successfully created edge types: [DRUGBANK_ddi_interactor_Compound2Compound].\n",
      "Successfully created edge types: [HETIONET_resembles_Compound2Compound].\n",
      "Successfully created edge types: [DRUGBANK_treats_Compound2Disease].\n",
      "Successfully created edge types: [GNBR_treatment_Compound2Disease].\n",
      "Successfully created edge types: [GNBR_inhibit_cell_growth_Compound2Disease].\n",
      "Successfully created edge types: [GNBR_sideEffect_adverseEvent_Compound2Disease].\n",
      "Successfully created edge types: [GNBR_alleviate_reduce_Compound2Disease].\n",
      "Successfully created edge types: [GNBR_disease_progress_biomarker_Compound2Disease].\n",
      "Successfully created edge types: [GNBR_prevent_suppress_Compound2Disease].\n",
      "Successfully created edge types: [GNBR_role_in_pathogenesis_Compound2Disease].\n",
      "Successfully created edge types: [HETIONET_treats_Compound2Disease].\n",
      "Successfully created edge types: [HETIONET_palliates_Compound2Disease].\n",
      "Successfully created edge types: [GNBR_improper_regulation_Gene2Disease].\n",
      "Successfully created edge types: [GNBR_causal_mutation_Gene2Disease].\n",
      "Successfully created edge types: [GNBR_polymorphisms_alter_risk_Gene2Disease].\n",
      "Successfully created edge types: [GNBR_role_in_pathogenesis_Gene2Disease].\n",
      "Successfully created edge types: [GNBR_possible_therapeutic_effect_Gene2Disease].\n",
      "Successfully created edge types: [GNBR_diognostic_biomarker_Gene2Disease].\n",
      "Successfully created edge types: [GNBR_promote_progression_Gene2Disease].\n",
      "Successfully created edge types: [GNBR_drug_targets_Gene2Disease].\n",
      "Successfully created edge types: [GNBR_overexpression_Gene2Disease].\n",
      "Successfully created edge types: [GNBR_mutations_affecting_disease_course_Gene2Disease].\n",
      "Successfully created edge types: [GNBR_in_tax_Gene2Tax].\n",
      "Successfully created edge types: [HETIONET_upregulates_Anatomy2Gene].\n",
      "Successfully created edge types: [HETIONET_expresses_Anatomy2Gene].\n",
      "Successfully created edge types: [HETIONET_downregulates_Anatomy2Gene].\n",
      "Successfully created edge types: [HETIONET_participates_Gene2Pathway].\n",
      "Successfully created edge types: [HETIONET_causes_Compound2Side_Effect].\n",
      "Successfully created edge types: [HETIONET_participates_Gene2Molecular_Function].\n",
      "Successfully created edge types: [HETIONET_participates_Gene2Biological_Process].\n",
      "Successfully created edge types: [HETIONET_participates_Gene2Cellular_Component].\n",
      "Successfully created edge types: [HETIONET_localizes_Disease2Anatomy].\n",
      "Successfully created edge types: [HETIONET_includes_Pharmacologic_Class2Compound].\n",
      "Successfully created edge types: [HETIONET_presents_Disease2Symptom].\n",
      "Successfully created edge types: [HETIONET_resembles_Disease2Disease].\n"
     ]
    }
   ],
   "source": [
    "print(conn.gsql('use global\\n' + schema))\n",
    "# print(conn.gsql('USE GLOBAL\\n DROP ALL')) # delete all the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Insert/Update the edge bulks chunk by chunk\n",
    "\n",
    "- Takes around 16 MINUTES! Be patient. (In my local TigerGraph, it takes around 60 seconds!)\n",
    "- Original data has 29 Genes without ids. I deleted them. See https://github.com/gnn4dr/DRKG/issues/32\n",
    "- The fixed drkg.tsv is [here](https://drive.google.com/file/d/1Hc1mMEyh_4p6qHm4VLK2isb-E51himCA/view?usp=sharing)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n"
     ]
    }
   ],
   "source": [
    "# Firstly, CREATE SCHEMA in your graph.\n",
    "# use edge_bulks to insert BIG data chunk by chunk\n",
    "CHUNK_SIZE = 100000\n",
    "cnt = 0\n",
    "prev_cnt_stop = 0 # if somehow an error occurs, set this variable to not start from zero \n",
    "for src_type in edge_bulks:\n",
    "    for edge in edge_bulks[src_type]:\n",
    "        for tgt_type in edge_bulks[src_type][edge]:\n",
    "            edges = edge_bulks[src_type][edge][tgt_type]\n",
    "            for i in range(0, len(edges), CHUNK_SIZE):\n",
    "                if cnt >= prev_cnt_stop: \n",
    "                    print(cnt)\n",
    "                    conn.upsertEdges(src_type, edge, tgt_type, edges[i:i + CHUNK_SIZE])\n",
    "                cnt += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n"
     ]
    }
   ],
   "source": [
    "# use node_bulks to insert BIG data chunk by chunk\n",
    "cnt = 0\n",
    "prev_cnt_stop = 0 # if somehow an error occurs, set this variable to not start from zero \n",
    "for node_type in node_bulks:\n",
    "    nodes = node_bulks[node_type]\n",
    "    for i in range(0, len(nodes), CHUNK_SIZE):\n",
    "        if cnt >= prev_cnt_stop: \n",
    "            print(cnt)\n",
    "            conn.upsertVertices(node_type, nodes[i:i + CHUNK_SIZE])\n",
    "        cnt += 1"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "c77c6f61c6b7b4b96b1f1362d1e492855f50ba30b6c9e892442a8cb8806ea707"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
